### PR TITLE
Feature - filter reports by agent, label and inboxes

### LINKED ISSUE #1218: Feature: Filter reports by agents
- Filter reports by agents

Acceptance criteria:

- Admin should be able to see reports of a particular agent.
- Admin should have a button to download a CSV report for  agents

## Design

Screenshot 2021-08-15 at 7 43 15 PM

### PR DESCRIPTION
**Pull request by fayazara**
## Description

Allows users to filter reports by agents, inboxes and labels.

https://user-images.githubusercontent.com/15716057/134824517-6d6e42b0-b921-457b-9432-77d949f8e947.mp4

Fixes 
#1218 
#1217
#2780 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/javascript/dashboard/assets/scss/widgets/_reports.scss [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  width: var(--space-two);
+}
+
+.reports-option__item {
+  flex-shrink: 0;
+  margin-right: var(--space-small);
+}
+
+.reports-option__label--swatch {
+  border: 1px solid #fff;
```
Comment (muhsin-k) [inline app/javascript/dashboard/assets/scss/widgets/_reports.scss:21]: Use CSS variable

File: app/javascript/dashboard/routes/dashboard/settings/reports/InboxReports.vue [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<template>
+  <woot-reports
+    key="inbox-reports"
+    type="inbox"
+    getter-key="inboxes/getInboxes"
+    action-key="inboxes/get"
+    :download-button-label="$t('REPORT.DOWNLOAD_AGENT_REPORTS')"
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/reports/InboxReports.vue:7]: Button text should be "Download inbox reports"

File: app/javascript/dashboard/routes/dashboard/settings/reports/LabelReports.vue [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<template>
+  <woot-reports
+    key="label-reports"
+    type="label"
+    getter-key="labels/getLabels"
+    action-key="labels/get"
+    :download-button-label="$t('REPORT.DOWNLOAD_AGENT_REPORTS')"
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/reports/LabelReports.vue:7]: Button text should be "Download label reports"

File: app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        :options="filterItemsList"
+        :option-height="24"
+        :show-labels="false"
+        @input="changeFilterSelection"
+      >
+        <template slot="singleLabel" slot-scope="props">
+          <img
+            class="reports-option__rounded--item reports-option__item"
+            :src="props.option.thumbnail"
+          />
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue:18]: Can we use the `Thumbnail` component here? https://github.com/chatwoot/chatwoot/blob/develop/app/javascript/dashboard/components/widgets/Thumbnail.vue

File: app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+          <span class="reports-option__desc">
+            <span class="reports-option__title">{{ props.option.name }}</span>
+          </span>
+        </template>
+        <template slot="option" slot-scope="props">
+          <div class="display-flex">
+            <img
+              class="reports-option__rounded--item reports-option__item"
+              :src="props.option.thumbnail"
+            />
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue:28]: Consider using  the component `Thumbnail` https://github.com/chatwoot/chatwoot/blob/develop/app/javascript/dashboard/components/widgets/Thumbnail.vue

**Review by muhsin-k — COMMENTED** [review]
@fayazara Nice work ❤️. Please see the inline comments

File: app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue (hunk 4 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue]: Fixed in https://github.com/chatwoot/chatwoot/pull/3084/commits/a80b6189e631d673b74c2dd13899346a8d4de25b

File: app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue (hunk 5 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue:21]: https://github.com/chatwoot/chatwoot/pull/3084/commits/a80b6189e631d673b74c2dd13899346a8d4de25b

File: app/javascript/dashboard/routes/dashboard/settings/reports/LabelReports.vue (hunk 3 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/settings/reports/LabelReports.vue]: Fixed in

https://github.com/chatwoot/chatwoot/pull/3084/commits/26ee689277da85b14e4706210103aef7bcb51d06

File: app/javascript/dashboard/routes/dashboard/settings/reports/InboxReports.vue (hunk 2 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/settings/reports/InboxReports.vue]: Fixed

https://github.com/chatwoot/chatwoot/pull/3084/commits/26ee689277da85b14e4706210103aef7bcb51d06

File: app/javascript/dashboard/assets/scss/widgets/_reports.scss (hunk 1 above)
Comment (fayazara) [reply app/javascript/dashboard/assets/scss/widgets/_reports.scss]: Fixed in https://github.com/chatwoot/chatwoot/pull/3084/commits/cb237fd3bd035954288c48f074e3308da2cb9815

**Review by nithindavid — CHANGES_REQUESTED** [review]

File: app/javascript/dashboard/api/reports.js [hunk 6]
```diff
@@ -6,15 +6,15 @@ class ReportsAPI extends ApiClient {
     super('reports', { accountScoped: true, apiVersion: 'v2' });
   }
 
-  getAccountReports(metric, since, until) {
+  getAccountReports(metric, since, until, type = 'account', id) {
```
[the remark this answers is not in the document]
Comment (fayazara) [reply app/javascript/dashboard/api/reports.js:9]: @nithindavid The `type` is a query parameter as mentioned in https://github.com/chatwoot/chatwoot/issues/2823.

File: app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue [hunk 7]
```diff
@@ -0,0 +1,209 @@
+<template>
+  <div class="flex-container flex-dir-column medium-flex-dir-row">
+    <div v-if="type === 'agent'" class="small-12 medium-3 pull-right">
+      <multiselect
```
[the remark this answers is not in the document]
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/settings/reports/components/ReportFilters.vue:7]: Resolved in https://github.com/chatwoot/chatwoot/pull/3084/commits/094283e5b8fbabc3b5e4b41a0ca9c48a8a6f2306

File: app/javascript/dashboard/helper/fileDownload.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+export default function downloadFile(fileName, content) {
```
Comment (muhsin-k) [inline app/javascript/dashboard/helper/fileDownload.js:1]: Add the specs for this helper

File: app/javascript/dashboard/api/reports.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (fayazara) [reply app/javascript/dashboard/api/reports.js]: Renamed methods in https://github.com/chatwoot/chatwoot/pull/3084/commits/a8f6ac32410baff93f698fd2cdaa6fedf9a35e3a

**Review by muhsin-k — DISMISSED** [review]

**Review by nithindavid — APPROVED** [review]

**Comment by pranavrajs** [comment]
@fayazara There are some broken tests in the frontend, can you please take a look?

**Review by pranavrajs — APPROVED** [review]

**Pull request merged**